src/diffusers/pipelines/marigold/pipeline_marigold_depth.py

```diff
@@ -0,0 +1,1246 @@
+        raise ValueError(f"Unexpected VAE input-output space size; {vae.config}")
+    if vae.config.latent_channels != 4:
+        raise ValueError(f"Unexpected VAE latent space size; {vae.config}")
+
+
+def find_batch_size(ensemble_size: int, resolution: int, dtype: torch.dtype) -> int:
+    bs_search_table = [
+        # tested on A100-PCIE-80GB
+        {"res": 768, "total_vram": 79, "bs": 36, "dtype": torch.float32},
+        {"res": 1024, "total_vram": 79, "bs": 20, "dtype": torch.float32},
+        # tested on A100-PCIE-40GB
+        {"res": 768, "total_vram": 39, "bs": 18, "dtype": torch.float32},
+        {"res": 1024, "total_vram": 39, "bs": 10, "dtype": torch.float32},
+        {"res": 768, "total_vram": 39, "bs": 30, "dtype": torch.float16},
+        {"res": 1024, "total_vram": 39, "bs": 15, "dtype": torch.float16},
+        # tested on RTX3090, RTX4090
+        {"res": 512, "total_vram": 23, "bs": 22, "dtype": torch.float32},
+        {"res": 768, "total_vram": 23, "bs": 9, "dtype": torch.float32},
+        {"res": 1024, "total_vram": 23, "bs": 5, "dtype": torch.float32},
+        {"res": 512, "total_vram": 23, "bs": 40, "dtype": torch.float16},
+        {"res": 768, "total_vram": 23, "bs": 18, "dtype": torch.float16},
+        {"res": 1024, "total_vram": 23, "bs": 10, "dtype": torch.float16},
+        # tested on GTX1080Ti
+        {"res": 512, "total_vram": 10, "bs": 6, "dtype": torch.float32},
+        {"res": 768, "total_vram": 10, "bs": 2, "dtype": torch.float32},
+        {"res": 1024, "total_vram": 10, "bs": 1, "dtype": torch.float32},
+        {"res": 512, "total_vram": 10, "bs": 10, "dtype": torch.float16},
+        {"res": 768, "total_vram": 10, "bs": 5, "dtype": torch.float16},
+        {"res": 1024, "total_vram": 10, "bs": 3, "dtype": torch.float16},
+    ]
+
+    if not torch.cuda.is_available():
+        logger.info("No GPU available, using batch_size=1")
+        return 1
+
+    total_vram = torch.cuda.mem_get_info()[1] / 1024.0**3
+    filtered_bs_search_table = [s for s in bs_search_table if s["dtype"] == dtype]
+    for settings in sorted(
+        filtered_bs_search_table,
+        key=lambda k: (k["res"], -k["total_vram"]),
+    ):
+        if resolution <= settings["res"] and total_vram >= settings["total_vram"]:
+            bs = settings["bs"]
+            if bs > ensemble_size:
+                bs = ensemble_size
+            elif (ensemble_size + 1) // 2 < bs < ensemble_size:
+                bs = (ensemble_size + 1) // 2
+            return bs
+
+    logger.info("Falling back to batch_size=1; feel free to set it manually to a higher value.")
+
+    return 1
+
+
+def ensemble_depth(
+    depth: torch.FloatTensor,
+    output_uncertainty: bool,
+    reduction: str = "median",
+    regularizer_strength: float = 0.02,
+    max_iter: int = 2,
+    tol: float = 1e-3,
+    max_res: int = 1024,
+) -> Tuple[torch.FloatTensor, Optional[torch.FloatTensor]]:
+    assert depth.dim() == 4 and depth.shape[1] == 1
```

Comment:
Can we go with explicit `raise ValueError()` everywhere instead of `assert`?

Reply:
I used the convention that anything user-facing will trigger a raise ValueError, and these asserts are more on the paranoid self-check side and must never be tripped (and if they ever are, that will be an indication that a raise ValueError must be added somewhere else). Let me know how to go about these, they can also be safely removed (although I usually keep them as the cost of those checks is insignificant)

Reply:
Now user-facing API raises, and any reusable member functions assert

Reply:
Is there any reason why shouldn't we always use `ValueError`? The user will still be getting an error without good explanations

Reply:
I use `assert` to perform sanity checks; it is there for me or other developers. `raise ValueError` is a similar but user-facing construction. Since I was suggested to remove `@torch.nograd` decorators from all methods except `__call__`, I figured that no other function is supposed to be called by a user. Do we want to have consistent `raise` checks regardless of how they are called? Or, if the other methods can also be user-facing, do we need to support `@torch.nograd` decorators on them?

Reply:
even though this is a private method and user cannot directly call it -  they can still pass the wrong input to the pipeline and trigger this assert error, no? otherwise why do we need it?

Reply:
You are right; I changed it to `raise ValueError` everywhere for consistency